src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs

```diff
@@ -0,0 +1,773 @@
+        {
+            var actionsList = new List<ICompareExchangeActions>();
+
+            try
+            {
+                foreach (var des in _destinations)
+                {
+                    actionsList.Add(des.CompareExchangeTombstones(_transactionOperationContext));
+                }
+                await foreach (var kvp in _source.GetCompareExchangeTombstonesAsync())
+                {
+                    var index = ShardHelper.GetShardIndexforDocument(_transactionOperationContext, _shardAllocation, kvp.Key.Key);
+
+
+                    await actionsList[index].WriteTombstoneKeyAsync(kvp.Key.Key);
+                }
+
+                return result.CompareExchangeTombstones;
+            }
+            finally
+            {
+                foreach (var action in actionsList)
+                {
+                    await action.DisposeAsync();
+                }
+            }
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessSubscriptionsAsync(SmugglerResult result)
+        {
+            await using(var actions = new SubscriptionActions(_server, _databaseRecord.DatabaseName))
+            {
+                return await ProcessSubscriptionsInternalAsync(result, actions);
+            }
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessReplicationHubCertificatesAsync(SmugglerResult result)
+        {
+            await using (var actions = new ReplicationHubCertificateActions(_server, _databaseRecord.DatabaseName))
+            {
+                return await ProcessReplicationHubCertificatesInternalAsync(result, actions);
+            }
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessTimeSeriesAsync(SmugglerResult result)
+        {
+            result.TimeSeries.Start();
+            try
+            {
+                for (int i = 0; i < _destinations.Count; i++)
+                {
+                    _actionsList.Add(_destinations[i].TimeSeries());
+                }
+                var isFullBackup = _source.GetSourceType() == SmugglerSourceType.FullExport;
+                await foreach (var ts in _source.GetTimeSeriesAsync(_options.Collections))
+                {
+                    HandleImportBatch("TimeSeries");
+                    var index = ShardHelper.GetShardIndexforDocument(_transactionOperationContext, _shardAllocation, ts.DocId);
+                    await ((ITimeSeriesActions)_actionsList[index]).WriteTimeSeriesAsync(ts);
+                    _itemsInBatch++;
+                }
+                return result.TimeSeries;
+            }
+            finally
+            {
+                await CleanActionList();
+            }
+        }
+
+        private async void HandleImportBatch(string type)
+        {
+            if (_itemsInBatch == 1000)
+            {
+                await SendBatch();
+
+                foreach (var des in _destinations)
```

Comment:
We have `foreach _destinations` in each and every one of the previous methods, than here as well.
_far_ better to have just one location for this.